evm/src/main/java/org/hyperledger/besu/evm/code/CodeV1.java

```diff
@@ -38,6 +573,216 @@ public class CodeV1 implements Code {
+        } else {
+          thisWork++;
+          continue;
+        }
+      }
+
+      while (currentPC < codeLength) {
+        int thisOp = code[currentPC] & 0xff;
+        int recordedStack = stackHeights[currentPC];
+        if (recordedStack >= 0 && currentStackHeight != recordedStack) {
+          return String.format(
+              "Stack height mismatch %d/%d at %d", recordedStack, currentStackHeight, currentPC);
+        }
+
+        byte[] stackInfo = opcodeStackValidation[thisOp];
+        int stackInputs;
+        int stackOutputs;
+        int pcAdvance = stackInfo[2];
+        if (pcAdvance == 0) {
+          return String.format("Invalid opcode 0x%02X", thisOp);
+        }
+        if (thisOp == CallFOperation.OPCODE || thisOp == JumpFOperation.OPCODE) {
+          int section = ((code[currentPC + 1] & 0xff) << 8) | (code[currentPC + 2] & 0xff);
+          stackInputs = codeSections[section].getInputs();
+          stackOutputs = codeSections[section].getOutputs();
+        } else {
+          stackInputs = stackInfo[0];
+          stackOutputs = stackInfo[1];
+        }
+
+        if (stackInputs > currentStackHeight) {
+          return String.format(
+              "Operation 0x%02X requires stack of %d but only has %d items",
+              thisOp, stackInputs, currentStackHeight);
+        }
+
+        currentStackHeight = currentStackHeight - stackInputs + stackOutputs;
+        if (currentStackHeight > MAX_STACK_HEIGHT) {
+          return "Stack height exceeds 1024";
+        }
+
+        maxStackHeight = Math.max(maxStackHeight, currentStackHeight);
+
+        if (thisOp == RelativeJumpOperation.OPCODE || thisOp == RelativeJumpIfOperation.OPCODE) {
+          // no `& 0xff` on high byte because this is one case we want sign extension
+          int rvalue = signedShortAt(code, currentPC);
+          workList[maxWork] = new int[] {currentPC + rvalue + 3, currentStackHeight};
+          maxWork++;
+        } else if (thisOp == RelativeJumpVectorOperation.OPCODE) {
+          int tableEnd = code[currentPC + 1] * 2 + currentPC + 2;
+          for (int i = currentPC + 2; i < tableEnd; i += 2) {
+            int rvalue = signedShortAt(code, i);
+            workList[maxWork] = new int[] {currentPC + rvalue + 3, currentStackHeight};
+            maxWork++;
+          }
+          currentPC = tableEnd - 2;
+        }
+        if (pcAdvance < 0) {
+          break;
+        }
+        currentPC += pcAdvance;
+        stackHeights[currentPC] = currentStackHeight;
+      }
+
+      thisWork++;
+    }
+
+    if (maxStackHeight != codeSections[codeSectionToValidate].maxStackHeight) {
+      return String.format(
+          "Calculated max stack height (%d) does not match reported stack height (%d)",
+          maxStackHeight, codeSections[codeSectionToValidate].maxStackHeight);
+    }
+    return null;
+  }
+
+  private static int signedShortAt(final byte[] code, final int currentPC) {
```

Comment:
Could migrate this to Words with his U16 friend. Also, had to read this twice, as it is not a signedShort at the PC, but rather right after it. Maybe `signedShortAfter` is a better name?

Reply:
Done, used I16 variant (U16 is unsigned)